File: 384.Shuffle-an-Array.go
// https://leetcode.com/problems/shuffle-an-array/
//
// algorithms
// Medium (48.9%)
// Total Accepted:    62,460
// Total Submissions: 127,700
// beats 90.48% of golang submissions

package leetcode

import "math/rand"

type Solution struct {
	initNums []int
}

func Constructor(nums []int) Solution {
	return Solution{
		initNums: nums,
	}
}

/** Resets the array to its original configuration and return it. */
func (this *Solution) Reset() []int {
	return this.initNums
}

/** Returns a random shuffling of the array. */
func (this *Solution) Shuffle() []int {
	length := len(this.initNums)
	shuffleNums := make([]int, length)
	copy(shuffleNums, this.initNums)

	for i := length - 1; i >= 0; i-- {
		idx := rand.Intn(i + 1)
		shuffleNums[i], shuffleNums[idx] = shuffleNums[idx], shuffleNums[i]
	}

	return shuffleNums
}

/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.Reset();
 * param_2 := obj.Shuffle();
 */

File: 384.Shuffle-an-Array.java
// https://leetcode.com/problems/shuffle-an-array/
//
// algorithms
// Medium (49.88%)
// Total Accepted:    74,127
// Total Submissions: 148,648


class Solution {

    private int[] nums;
        
    public Solution(int[] nums) {
        this.nums = nums;
    }
    
    /** Resets the array to its original configuration and return it. */
    public int[] reset() {
        return nums;
    }
    
    /** Returns a random shuffling of the array. */
    public int[] shuffle() {
        int[] curNums = this.nums.clone();
        int length = curNums.length;
        int tail = length;
        
        Random random = new Random();
        
        for (int i = 0; i < length - 1; i++) {
            int idx = random.nextInt(tail);
            swap(curNums, idx, tail - 1);
            tail--;
        }
        
        return curNums;
    }
    
    public void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int[] param_1 = obj.reset();
 * int[] param_2 = obj.shuffle();
 */